Add all declared scopes is auth.oauth2.scopes discovery file section

clementdenis opened the pull request:
Currently, only the email scope is asked by the API explorer, because other scopes declared by APIs and API methods are not aggregated in the discovery file.
This brings Cloud Endpoints on par with the behavior of Google own APIs.

Commit:
Add all declared scopes is auth.oauth2.scopes discovery file section

tangiel commented on endpoints-framework/src/main/resources/scopeDescriptions.properties:
```diff
@@ -0,0 +1,219 @@
+#Source: https://developers.google.com/identity/protocols/googlescopes
```
Is this file generated or hand written? Might be a hassle to keep up to date.

clementdenis replied:
Keys and values were extracted separately using the xpath expressions below, and merged in a spreadsheet.
Then some transformations were performed to deduplicate, sort the entries and add comments.

All of this could be easily automated, but I was hoping you had an "official" list of scope descriptions at Google.
This would be much better than scrapping a web page. Do you know if something like that exists?

tangiel reviewed (commented):
Apologies for the slow review.

Commit:
Add AuthScopeDescriptions to fetch updated scopes automatically
Move scopeDescriptions.properties to DiscoveryGenerator packge

clementdenis commented:
Hi @tangiel ,
Could you have another look at this one?
I added AuthScopeDescriptions to fetch up-to-date scopes automatically (updating scopeDescriptions.properties is still manual though).

tangiel commented on endpoints-framework/src/main/java/com/google/api/server/spi/discovery/DiscoveryGenerator.java:
```diff
@@ -174,13 +193,21 @@ private RestDescription writeApi(ApiKey apiKey, Iterable<ApiConfig> apiConfigs,
     }
 
+    LinkedHashMap<String, ScopesElement> scopeElements = new LinkedHashMap<>();
+    for (String scope : allScopes) {
```
Presumably this doesn't work when multiple scopes are required? What would happen if the scope expression is "email profile" for example?

clementdenis replied:
Indeed, but is this how the @Api*.scopes field is supposed to be used? 
It is a String array, so I assumed each scope must be declared independently.
i.e. @ApiMethod(scopes = {"email", "profile"}) rather than @ApiMethod(scopes = {"email profile"})

tangiel replied:
This probably needs to be documented better, but the comma separation is a logical OR for scopes. To require multiple scopes simultaneously you are supposed to use one string with space separation.

Commit:
Extract auth scope description logic to its own class
- follows the schema repository pattern already used in DiscoveryGenerator
- could be reused for swagger generation (if / when supporting accessToken security definition)
- added some tests for proper merging of existing scope declarations

clementdenis commented:
I extracted the logic for scope description handling to its own class, should be clearer.

tangiel commented:
I don't know if you saw my comment since it was in an outdated comment, but: This probably needs to be documented better, but the comma separation is a logical OR for scopes. To require multiple scopes simultaneously you are supposed to use one string with space separation.

Commit:
Handle scope conjunctions properly

tangiel commented:
Thanks @clementdenis, sorry it took so long to review.
